### PR TITLE
Models refactor

### PR DESCRIPTION
**Pull request by mmuckley**
This PR refactors the models of NeuralCompression. At a high level it does two things:

1. We remove all duplicate model implementations that are currently implemented in CompressAI.
2. We add an implementation of HiFiC, which will be later used for the downstream open-sourcing of MS-ILLM.

As we plan to use several variants of the Hyperprior model, we do introduce a new abstract base class for hyperprior models that encapsulates the HiFiC autoencoder. This will make it easier to build a common interface for training loops involving these classes of models.

The test suite has also been updated to test the new HiFiC autoencoder and its functionality, as well as removing old tests that references deleted code.

### CODE REVIEW

File: neuralcompression/functional/_pad_image.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    Returns:
+        The image padded so that its dimensions are disible by factor, as well
+        as the height and width.
+    """
+    # pad image if it's not divisible by downsamples
+    _, _, height, width = image.shape
+    pad_height = (factor - (height % factor)) % factor
+    pad_width = (factor - (width % factor)) % factor
+    if pad_height != 0 or pad_width != 0:
+        image = F.pad(image, (0, pad_width, 0, pad_height), mode="reflect")
```
Comment (QuentinDuval) [inline neuralcompression/functional/_pad_image.py:33]: You could make the `mode` part of the arguments taken by the function (and make it default to "reflect")

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+
+import pytest
+import torch
+
+
+@pytest.fixture(
+    scope="session", params=[(2, 2, 64, 64), (3, 5, 128, 64), (1, 3, 256, 256)]
+)
+def arange_4d_image(request):
```
Comment (QuentinDuval) [inline tests/conftest.py:14]: I don't get which function this test is suppose to cover

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    x = torch.arange(torch.prod(torch.tensor(request.param))).reshape(request.param)
+
+    return x.to(torch.get_default_dtype())
+
+
+@pytest.fixture(
+    scope="session", params=[(2, 2, 65, 257), (3, 5, 124, 63), (1, 3, 252, 257)]
+)
+def arange_4d_image_odd(request):
```
Comment (QuentinDuval) [inline tests/conftest.py:23]: Same for this test

File: tests/layers/test_channelnorm.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import pytest
+import torch
+from torch import Tensor
+
+from neuralcompression.layers import ChannelNorm2D
+
+
+@pytest.mark.parametrize("affine", [True, False])
+def test_channel_norm_2d(affine, arange_4d_image: Tensor):
```
Comment (QuentinDuval) [inline tests/layers/test_channelnorm.py:14]: This type of test is not my favorite: the test verifies that the content of the module is the same as the content of the module (since the test replicates the logic of the module).

I would try a more realistic tests: iterate several time on the same input and see that it does normalize ? Or just drop the test.

**Review by QuentinDuval — APPROVED** [review]

File: tests/conftest.py (hunk 3 above)
Comment (mmuckley) [reply tests/conftest.py:23]: See comment for `arange_4d_image`.

File: tests/functional/test_pad_image.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import pytest
+from torch import Tensor
+
+import neuralcompression.functional as ncF
+
+
+@pytest.mark.parametrize("factor", [64, 32, 16])
```
[the remark this answers is not in the document]
Comment (mmuckley) [reply tests/functional/test_pad_image.py:12]: Makes sense to me. I've made this change.

File: tests/layers/test_channelnorm.py (hunk 4 above)
Comment (mmuckley) [reply tests/layers/test_channelnorm.py:14]: Yeah I didn't like this a ton - but it's more to guard against an implementation change of the module in the future.

Running multiple times won't change anything (unless you want to verify that the mean is still 0 and the variance is still 1). We can think a bit about it.

**Comment by mmuckley** [comment]
Merging after offline discussion with reviewer.
